File: app/src/main/java/ch/bailu/aat/dispatcher/BeaconSource.java
package ch.bailu.aat.dispatcher;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import ch.bailu.aat.gpx.GpxInformation;
import ch.bailu.aat.gpx.InfoID;
import ch.bailu.aat.util.AppBroadcaster;
import ch.bailu.aat.services.ServiceContext;

public final class BeaconSource extends ContentSource {
    private final ServiceContext scontext;

    public BeaconSource (ServiceContext sc) {
        scontext = sc;
    }

    private final BroadcastReceiver onTrackChanged = new BroadcastReceiver () {
        @Override
        public void onReceive(Context context, Intent intent) {
            sendUpdate(InfoID.BEACON, scontext.getBeaconService().getLoggerInformation());
        }
    };

    @Override
    public void requestUpdate() {
        sendUpdate(InfoID.BEACON, scontext.getBeaconService().getLoggerInformation());
    }

    @Override
    public void onPause() {
        scontext.getContext().unregisterReceiver(onTrackChanged);
    }

    @Override
    public void onResume() {
        AppBroadcaster.register(scontext.getContext(), onTrackChanged, AppBroadcaster.BEACON);
    }

    @Override
    public int getIID() {
        return InfoID.BEACON;
    }

    @Override
    public GpxInformation getInfo() {
        return scontext.getBeaconService().getLoggerInformation();
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/source/BeaconSource.kt
package ch.bailu.aat_lib.dispatcher.source

import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.broadcaster.BroadcastReceiver
import ch.bailu.aat_lib.broadcaster.Broadcaster
import ch.bailu.aat_lib.dispatcher.SourceInterface
import ch.bailu.aat_lib.dispatcher.TargetInterface
import ch.bailu.aat_lib.dispatcher.usage.UsageTrackerInterface
import ch.bailu.aat_lib.gpx.information.GpxInformation
import ch.bailu.aat_lib.gpx.information.InfoID
import ch.bailu.aat_lib.service.ServicesInterface

class BeaconSource(
    private val services: ServicesInterface,
    private val broadcaster: Broadcaster,
    usageTracker: UsageTrackerInterface
) : SourceInterface {
    private var target = TargetInterface.NULL

    init {
        usageTracker.observe {
            requestUpdate()
        }
    }

    override fun setTarget(target: TargetInterface) {
        this.target = target
    }

    private val onTrackChanged = BroadcastReceiver {
        requestUpdate()
    }

    override fun requestUpdate() {
        target.onContentUpdated(InfoID.BEACON, services.getBeaconService().getLoggerInformation())
    }

    override fun onPauseWithService() {
        broadcaster.unregister(onTrackChanged)
    }

    override fun onDestroy() {}

    override fun onResumeWithService() {
        broadcaster.register(AppBroadcaster.BEACON, onTrackChanged)
    }

    override fun getIID(): Int {
        return InfoID.BEACON
    }

    override fun getInfo(): GpxInformation {
        return services.getBeaconService().getLoggerInformation()
    }
}